### ISSUE TITLE
[1.16.5] Client side PartEntity id doesn't match with server side

### ISSUE THREAD

**Issue by XenonDB** [MinecraftForge/MinecraftForge#8188]
**Minecraft Version:** 1.16.5

**Forge Version:** 1.16.5-36.1.0 & 1.16.5-36.2.9

**Description of issue:**

I'm making a system to let most entity adapted with PartEntity system of forge.
However, logs show the id of PartEntity from ender dragon in client side doesn't match with server side (see logs below)
These are generated by mixin in forge-1.16.5-36.1.0.

I think the reason is here:
https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/patches/minecraft/net/minecraft/client/network/play/ClientPlayNetHandler.java.patch#L49

i starts from 0 and causes id shifting (first PartEntity's id is same as its parent. However each entity should have different id)

This seems still exist in forge-1.16.5-36.2.9:
![未命名](https://user-images.githubusercontent.com/58476103/140300106-d7d861af-f693-475c-baeb-a2f5b736d21b.png)

**Logs:** 
![unknown](https://user-images.githubusercontent.com/58476103/140298465-392003ad-9782-42e6-8dd9-62f1c06ca12d.png)

is this a vanilla or forge bug?

**Comment by uSkizzik** [comment]
This is more of a vanilla bug since the linked line isn't changed by Forge.
Forge Part Entities also **do not** touch IDs, so they can't be causing the issue.

**Comment by ChampionAsh5357** [comment]
This is already confirmed and still unresolved in the Jira [MC-158205](https://bugs.mojang.com/browse/MC-158205).

**Comment by uSkizzik** [comment]
Actually, I experienced this exact thing too! One of my entity's parts wasn't taking any damage. I tried fixing it but I gave up and left it like it was.

**Comment by XenonDB** [comment]
ok...maybe I need to fix it by myself

**Comment by uSkizzik** [comment]
> ok...maybe I need to fix it by myself

PR?

**Comment by XenonDB** [comment]
> > ok...maybe I need to fix it by myself
> 
> PR?

Not sure if forge want to fix vanilla bug.
Maybe I will use mixin or prevent my custom PartEntity to be set a wrong id

**Comment by ChampionAsh5357** [comment]
Why would you need a mixin for your own part entity? Just add 1 to the incoming client part entities and that should suffice.

**Comment by uSkizzik** [comment]
> Not sure if forge want to fix vanilla bug.
It would be beneficial to everyone who uses the Part Entity system, so I don't see why not.

**Comment by XenonDB** [comment]
> Why would you need a mixin for your own part entity? Just add 1 to the incoming client part entities and that should suffice.

I mean use mixin on ClientNetPlayHandler.

And yes, making a pull request also maybe a good idea.

**Comment by uSkizzik** [comment]
> And yes, making a pull request also maybe a good idea.

We actually talked about this in Discord and turns out the fix applies per entity.
Turns out the fix is rly simple. You just need to add one when setting the id of your part. (In the recreateFromPacket method or whatever it's called in MCP mappings)
This is what I'm talking about:
```java
for(int i = 0; i < parts.length; ++i) {
  parts[i].setId(i + 1 + packet.getId());
}
```

**Comment by XenonDB** [comment]
ok I just made a pull request

**Comment by Insane96** [comment]
Actually the wings bug is this report [MC-225055](https://bugs.mojang.com/browse/MC-158205). As I wrote in [MC-158205](https://bugs.mojang.com/browse/MC-158205):

```Seems like they're (wings ID) swapped between client and server. So if right wing has Entity ID 1 and left wing has Entity ID 2 on the client, for the server it'll be Entity ID 2 for the right wing and Entity ID 1 for the left wing. With this the server thinks that you're hitting the far away wing so since there's a max range for the hit of 6 blocks you can't reach it and so the hit is registered on the client (MC-225055) but not on the server.```

But this issue doesn't show that they're swapped, instead there's a shift in the EntityID.

**Comment by ChampionAsh5357** [comment]
The wings bug is only one part of the problem and the issue relates to all collisions boxes and not specifically 225055 as that is just a result of 158205 failing. If you could provide the research for how you game to the swapping conclusion, that would give a better sense whether there is more than the above issue at work here.

**Comment by Insane96** [comment]
This Forge issue relates to MC-225055, which makes the player fail to hit wings in almost any case (unless the player's at least 6 blocks from the server side's Part ID, which is not swapped as I wrote in the comment, but instead is shifted).

MC-158205 is, instead, the issue where the dragon entity is in a chunk that the player doesn't load when attacking.

They're two different issues.

**Comment by ChampionAsh5357** [comment]
Again, how did you come to this conclusion? I would like to see your reasoning because the backing logic you are describing does not seem present.

Chunks are loaded in a 11x11 radius around all players. The entity hit result gets the closest entity box to the player that is not in spectator and is pickable, both of which are satisfied. The parts of the entities are also explicitly checked to add all parts if the entity itself is selectable. It checks all loaded chunks that the bounding box intersects (which in the case of vanilla melee attacks will always be loaded). Since the dragon is a 16x16x8, it will be able to be checked assuming the player is within range. It then checks each part individually to see if the raycast was able to hit. 

The interesting thing here is that the entire span of the dragon from wing to wing is 14 blocks, from head to tail is 15 blocks, and from height 3 blocks at most. This means it's quite difficult to hit the dragon with a melee attack unless you are almost directly within the entity bounding box. Couple that with the desynced entity part ids and you get a very hard to hit dragon with a melee attack. Even more so, there is a bit of extra processing where the main entity bounding box is not considered in damageability and that any attack not to the head is divided by 4 and then adds 1 to damage in most cases makes a difficult box.

See `Level#getEntities` which is called from `ProjectileUtil#getEntityHitResult` used to handle this logic. You can also see `Minecraft#startAttack` to view that the `hitResult` is called which is checked and updated every tick within `Minecraft#tick` along with every frame within `GameRenderer#renderLevel`.

**Comment by Insane96** [comment]
No, when hitting an entity the game checks the dragon entity not the single
part, this means that the dragon will not be check if not in the chunks of
the raycast.

**Comment by ChampionAsh5357** [comment]
Again, can you please show me where in the code that logic exists? I have given you exactly where the logic I am referring to lies which, as far as I know and that no one has contested, proves my statement. As mentioned before, if the bounding box of the dragon is within the raycast, all parts of the dragon are checked as well regardless of if they were part of the original raycast. I linked you the location of this logic within `Level#getEntities`. So please show where you are making your argument from and on what basis of evidence.

**Comment by Insane96** [comment]
The problem lies client side where `GameRenderer#getMouseOver` (mcp mappings) cannot set the `Minecraft#objectMouseOver` to the entity as the raycast fails.

In the screenshot I'm actually hovering the ender dragon head and/or neck and the breakpoint is not hit as the raycast fails.
![image](https://user-images.githubusercontent.com/7185104/140523885-39585fbe-4f46-443f-a3b1-f8f67dbe6a53.png)

This results in `Minecraft#clickMouse` to fail sending the hit to the server.

**Comment by ChampionAsh5357** [comment]
I'm confused how that relates to the chunk however since that's literally how the entity is selected. This does not really show anything but the place you are clicking is not hitting the entity box which could be true, though you could be too far away or the boxes are not in the places you are hovering over. The code I'm referring to immediately proceeds the breakpoint you have set.

Here is a full test where we can see what the hit result is checking to verify that we are targeting the correct part to verify. Also, let's use the dragon's sitting position on the pedestal as our test case since the client and server side positions will be properly synced to ensure that this is not because of some desync situation. Let's modify the debug code to return the entity that the player is hovering over instead of the the crosshair entity:
```java
// Within DebugOverlayGui#getSystemInformation (1.16 MCP Classnames, Mojmap Fields/Methods)
Entity entity = this.minecraft.hitResult instanceof EntityRayTraceResult ? ((EntityRayTraceResult) this.minecraft.hitResult).getEntity() : null;
if (entity != null) {
   list.add("");
   list.add(TextFormatting.UNDERLINE + "Targeted Entity");
   list.add(entity.getClass().getSimpleName() + "['" + entity.getName().getString() + "'/"  + entity.getId() + "]");
   if (entity instanceof EnderDragonPartEntity) list.add(((EnderDragonPartEntity) entity).name);
   //list.add(String.valueOf((Object)Registry.ENTITY_TYPE.getKey(entity.getType())));
   entity.getType().getTags().forEach(t -> list.add("#" + t));
}
```

Now, let's take another look. Let's try targeting the head or neck:
![Head Target No Id Changes](https://i.imgur.com/OT8mIM7.png)
Well it seems targeting the head doesn't damage the entity. Though that's to be expected. It has the same id as the main dragon so it will be treated as if it was the actual dragon which cannot take damage.
![Neck Target No Id Changes](https://i.imgur.com/UmvKM3Y.png)
Targeting the neck and attacking is doing damage, which is expected, though it would be doing so of the head's damage instead because of the id shift.

There will be no error thrown as well since the ids are shifted one to the right on the client. This means that every other part will have an id one less of that of its serverside counterpart never resulting in a null id.

Let's try adjusting the id offsets to the correct ids now and see what that does:
```java
// Within ClientPlayNetHandler#handleAddMob  (1.16 MCP Classnames, Mojmap Fields/Methods)
if (livingentity instanceof EnderDragonEntity) {
   EnderDragonPartEntity[] aenderdragonpartentity = ((EnderDragonEntity)livingentity).getSubEntities();

   for(int i = 0; i < aenderdragonpartentity.length; ++i) {
      aenderdragonpartentity[i].setId(i + 1 + p_147281_1_.getId());
      // aenderdragonpartentity[i].setId(i + p_147281_1_.getId());
   }
}
```

Let's try targeting the head or neck once more:
![Head Target With Changes](https://i.imgur.com/pjumMLr.png)
The head can now be targeted correctly!
![Neck Target With Changes](https://i.imgur.com/Nsvmm5r.png)
And the neck still works as well.

### Further Remarks

Notice that using the above debug code that the Ender Dragon is never shown in the debug menu. This is because `Entity#isPickable` returns false which means that the dragon itself can never be picked as `ProjectileHelper#getEntityHitResult` is supplied a predicate from `GameRenderer#pick` where `Entity#isSpectator` is false and `Entity#isPickable` is true. To make sure that parts could still be selected even if the parent couldn't be, the check for the parts is first surrounded by an intersection of the main entity. Then it checks whether there is a predicate and the main entity returns true **separately** from the entity parts. Then it checks whether the entity parts intersect and match the predicate. Here is a snippet from `Chunk#getEntities` to show this:
```java
if (entity.getBoundingBox().intersects(boxBeingChecked) && entity != excludedEntity) {
   if (canEntityBeSelected == null || canEntityBeSelected.test(entity)) {
      selectableEntities.add(entity);
   }
   if (entity instanceof EnderDragonEntity) {
      for(EnderDragonPartEntity enderdragonpartentity : ((EnderDragonEntity)entity).getSubEntities()) {
         if (enderdragonpartentity != excludedEntity && enderdragonpartentity.getBoundingBox().intersects(boxBeingChecked) && (canEntityBeSelected == null || canEntityBeSelected.test(enderdragonpartentity))) {
            selectableEntities.add(enderdragonpartentity);
         }
      }
   }
}
```

This is the reasoning I described above. Feel free to use the above code to verify my test cases and logic reasoning.

**Comment by XenonDB** [comment]
Here is one of my testing report in forge-1.17.1-37.0.104
And, interesting, sometimes the parts id don't seem continuous on server:
![wtf](https://user-images.githubusercontent.com/58476103/140566935-28029729-9040-42f4-a0fd-959b23d5084a.png)
```java
@SubscribeEvent
    public void testEnderDragon(AttackEntityEvent event) {
    	
    	Entity e = event.getTarget();
    	
    	LOGGER.info("------------------");
    	
    	LOGGER.info(e.toString());
    	LOGGER.info(String.format("target id: %d", e.getId()));
    	
        if(e instanceof EnderDragonPart) {
        	
        	EnderDragon d = ((EnderDragonPart) e).getParent();
        	
        	LOGGER.info(d.toString());
        	LOGGER.info(String.format("dragon id: %d", d.getId()));
        	
        	EnderDragonPart[] parts = (EnderDragonPart[]) d.getParts();
        	
        	for(EnderDragonPart part : parts) {
        		LOGGER.info(part.toString());
            	LOGGER.info(String.format("dragon part: %s , id: %d", part.name,part.getId()));
        	}
        	
        }
        
        LOGGER.info("------------------");
    }
```
The id is continuous most of the time.

Still don't know how this happened.
This may mean client side can't predict ids of parts on server side just according to dragon.

**Comment by ChampionAsh5357** [comment]
That would be another issue then since that would mean the client would be even more incorrectly displaying the associated data. ~~I think this is a bug due to how entities can now be spawned in an asynchronous fashion for chunk loading.~~ Shouldn't say things unless I can verify it.

**Comment by XenonDB** [comment]
Asynchronous chunk loading?...oh no.
Parts' id are continuous because these are creating in constructor of ender dragon.
If another thread can preempt a thread which is creating an dragon, it can't guarantee parts' id are continuous and need another way to tell client what part id is or need a lock for entity id.

**Comment by XenonDB** [comment]
> I think this is a bug due to how entities can now be spawned in an asynchronous fashion for chunk loading.

Is this a vanilla feature?

**Comment by sciwhiz12** [comment]
Seems this has been fixed by #8576, published as `1.18.2-40.1.10`. For 1.16, an LTS backport PR is open at #8605. I'll be closing this issue now as 1.18 has a fix and 1.16 has a fix under PR review.
